### PR TITLE
update steel/ord

### PR DESCRIPTION
**Pull request by fominok**
Hey,
I'm not sure what is the goal regarding consistency with Scheme, but according to Racket docs for `>`, `<`, `>=` and `<=` functions both args shall be `real?`. Also things like `(< 1337 "leet")` didn't fail either.

This PR updates `steel/ord` module for real numbers comparison as well as adds missing documentation.

P.S. I tried to implement something like numeric tower as well to promote types in some more typesafe way to possibly reuse it in other modules, but failed miserably, shall be a separate effort I suppose, going the most direct way now.

### CODE REVIEW

**Comment by fominok** [comment]
Unfortunately I can't run all tests locally to see what's going on, but my not so wild guess is that ord tests exist for strings and for numbers and both use same functions, so fix for strings to use char comparison functions breaks the same test that is fed with numbers; i don't know what the final decision here, shall `>` be generic as possible or we still take inspiration from Racket?

**Comment by mattwparas** [comment]
To answer the notion that `<` should only work with numbers, I'm inclined to agree with that sentiment, matching racket and guile would be a good idea then

**Comment by mattwparas** [comment]
Also, running the tests locally, seems like just a mixup - you did edit the `trie.rkt` but the test is running a different `trie` file - just update `cogs/sorting/trie-sort.scm` to with the `char<?` and the tests will pass

**Comment by fominok** [comment]
> Also, running the tests locally, seems like just a mixup - you did edit the `trie.rkt` but the test is running a different `trie` file - just update `cogs/sorting/trie-sort.scm` to with the `char<?` and the tests will pass

it seems there are two trie-sort.scm in tests and I fixed only one besides trie.rkt, thanks

**Comment by fominok** [comment]
Since it's passing, does it require some extra work to be merged?

**Comment by mattwparas** [comment]
Sorry! Been a very busy few days. I will review it today

File: crates/steel-core/src/steel_vm/primitives.rs
Comment (mattwparas) [inline crates/steel-core/src/steel_vm/primitives.rs:960]: Not sure if you saw this, but this is how `PartialOrd` is implemented for steel values as of now: https://github.com/mattwparas/steel/blob/master/crates/steel-core/src/rvals.rs#L2081

Could we just lift the comparisons for numbers and re use them here?

Comment (fominok) [reply crates/steel-core/src/steel_vm/primitives.rs:960]: Indeed I didn't see it, thanks

Did I understand correctly that you suggest to improve those existing ones?

Comment (mattwparas) [reply crates/steel-core/src/steel_vm/primitives.rs:960]: Sorry for the delay, been a busy few days. I didn't do an exhaustive comparison, but mostly suggesting using the existing comparison code as reference to make sure we're not missing anything - whether that means improving what is there or for the stuff you've written. I'll do another pass shortly to see what the differences are if any

Comment (fominok) [reply crates/steel-core/src/steel_vm/primitives.rs]: Hey, i've reused existing PartialEq since it's there, and updated it as well in order not to loose precision easily, per TODO above

File: crates/steel-core/Cargo.toml
Comment (mattwparas) [inline crates/steel-core/Cargo.toml:71]: Is this additional dependency needed?

Comment (fominok) [reply crates/steel-core/Cargo.toml:71]: I used bigdecimals to cover cases with float and something possibly bigger than f64, like BigRational

Comment (mattwparas) [reply crates/steel-core/Cargo.toml:71]: I'm not explicitly against bringing in another dependency, but what was the behavior before? Were we just losing precision then? I'm not sure if scheme even has big floats as a data type in general so I'm not explicitly sure if this is necessary

Comment (fominok) [reply crates/steel-core/Cargo.toml:71]: Yes, precision, and need to manually handle Big* types conversion if they can't fit in f64 if we do without BigDecimal, to shortcut to "greater than"

Comment (mattwparas) [reply crates/steel-core/Cargo.toml:71]: Got it, I think this is fine then. Sorry for the delay once again

File: crates/steel-core/src/steel_vm/primitives.rs
Comment (mattwparas) [inline crates/steel-core/src/steel_vm/primitives.rs:975]: So the only other comment I'd have is that these functions are typically multi-arity - as in they take an arbitrary number of parameters. Could we use the rest args here to implement that? That would apply to all of the primitives here as well

File: crates/steel-core/Cargo.toml
Comment (fominok) [reply crates/steel-core/Cargo.toml:71]: No worries, thanks!

File: crates/steel-core/src/steel_vm/primitives.rs
Comment (fominok) [reply crates/steel-core/src/steel_vm/primitives.rs]: Hey, I've updated the PR; One thing I noticed though is that `arty = "AtLeast(1)"` seems to do nothing and proceeds with things like `(>)`

Comment (mattwparas) [reply crates/steel-core/src/steel_vm/primitives.rs]: I believe the arity annotation only comes into play with documentation, so it won't enforce any checks

Comment (mattwparas) [reply crates/steel-core/src/steel_vm/primitives.rs]: So in this case - the only thing remaining is to add an arity check that we're passing at least one argument in (which I think is the behavior we want) - otherwise this LGTM

Comment (fominok) [reply crates/steel-core/src/steel_vm/primitives.rs]: Got it, updated the check

Then I think we're good to go

**Review by mattwparas — APPROVED** [review]

**Pull request merged**
